### ISSUE TITLE
scattering environments as parameters

### ISSUE THREAD

**Issue by satra** [common-workflow-language/common-workflow-language#135]
`use case:`

for many settings we often want to ensure that the same workflow runs under different environments and versions of software. one way is to build docker containers for each version (a la travis). however, in many practical cases we may want to run such regression testing on our own machines and clusters.

we have been working on an extension of `unittest` called [testkraut](https://testkraut.readthedocs.org/en/latest/) that we were considering moving over to CWL and spent some time hacking and understanding how we could.

`current hack:`

create a wrapper cwl that takes as an input an environment spec/docker image id, then the internal tool writes a cwl file with the environment or requirements based on the parameter and runs the workflow.

**Comment by tetron** [comment]
It sounds like you want to parameterize the process requirements.  You can already do this for environment variables and configuration files. You do need to translate each environment into a CWL description, as CWL discourages the "run a magic bash script" approach to configuring the environment.  Does this make sense?  If you can describe a specific case of something you do now I can provide an example in CWL and see if it meets your needs.

**Comment by satra** [comment]
@tetron - actually both travis and circleci specs kind of cover this idea. for example in travis we can do something like a testing matrix.

http://docs.travis-ci.com/user/customizing-the-build/#Build-Matrix

**Comment by satra** [comment]
@tetron -  as a follow up to this:

> You do need to translate each environment into a CWL description, as CWL discourages the "run a magic bash script" approach to configuring the environment. 

could you please clarify this?

i can't see how you will avoid this without restricting folks to use docker images and vms. 

for example, even installing a python environment through say conda - from continuum.io - would require a sequence of statements. and since package management across tools and environments is not a solved issue, it seems that the only way out is to use images, which would have required some script (bash, fabric, ansible, dockerfile, etc.,.) to configure.

**Comment by tetron** [comment]
I see, I think we're talking about slightly different things.

What I was talking about is that to configure a tool to use an existing, installed environment, you should use CWL `EnvVarRequirement` or `CreateFileRequirement` to set the specific environment variables or configuration files the tool needs.  The "magic bash script" I'm referring to is the typical environment setup script that mostly just sets environment variables (sometimes with some noisy conditional logic in an attempt to make it portable).  For example virtualenv "activate": despite being 80 lines long all it does that actually matters for running Python programs is set `VIRTUAL_ENV` and `PATH`.

Aside from providing Docker images or Dockerfiles, CWL currently doesn't say anything about installing an environment (or detecting an existing environment).  We've been mulling over the idea of generalizing the notion of a container to include approaches other than Docker in issue #80.

**Comment by mr-c** [comment]
@satra Thank you for starting this conversation, have we addressed your issue?

**Comment by satra** [comment]
@mr-c - i don't think this is resolved fully. let me restate my requirement as the following:

travis is a workflow. how would you recreate the travis workflow using cwl? perhaps that sort of general thinking is beyond the scope of cwl, but to be common, i think it should be.

from a spec perspective, one should be able to take any line that is an "input" and scatter it, and then collate it. right now what can and cannot be scattered is compartmentalized to "inputs" to a tool, not how the tool runs, etc.,. therefore one can hack cwl to kind of do what i want. i.e. write a tool that takes an environment configuration (say py3 vs py2, numpy 1.x vs 1.y, glibc 2.XX) and either crashes because it can't create that environment, or having satisfied the requirements, instantiates or ensures the environment and runs some underlying script within the environment.

in general this is hard, which is precisely why docker/vagrant are good intermediate options. but even running things within different docker environments would be a hack. so from a more general perspective i would like to scattering be applied to any leaf of the cwl spec tree.

**Comment by ntijanic** [comment]
> so from a more general perspective i would like to scattering be applied to any leaf of the cwl spec tree.

A bit meta and hacky, but there is a way to do this with CWL:

- Describe desired tool/workflow in CWL, without specifying environment (`tool`).
- Create a tool that merges CWL Requirement(s) with CWL file and describe it in CWL (`merge`).
- Describe cwltool in CWL (`cwltool`).
- Specify inputs to desired tool/workflow as a single JSON object (`tool_inputs`).
- Specify desired environments as a list of CWL Requirement objects (`env_requirements`).
- Create the following CWL workflow: `[cwltool(tool_with_env, tool_inputs) for tool_with_env in [merge(tool, env) for env in env_requirements]]`

Sadly, that would still require creating all combinations of requirements. This can also be worked around by making multiple Dockerfiles in a step in the outer workflow or by creating a wrapper around cwltool that includes setting up the environment based on one of the inputs.

Frankly, this is probably better solved with a script that submits jobs based on a template and different runtime requirements. I don't think scattering on _any_ CWL node is feasible, but having processes/steps with some control inputs/outputs (on which one can do scatter) sounds interesting.

**Comment by mr-c** [comment]
@satra I agree that it would be useful to scatter over environment requirements, just like any other input.

`cwltool` now supports overriding requirements in the input object, so you can address this without writing new CWL descriptions by varying the input document itself. However this wasn't documented in the 1.0 specification; it will be in the next release via #70
